### ISSUE TITLE
More Flow Operators

### ISSUE THREAD

**Issue by Neverlord** [actor-framework/actor-framework#1389]
CAF 0.19 introduced ReactiveX-style flow processing with a base set of the most useful operators. There are more operators defined by ReactiveX and eventually we should implement all of them:

## Creating Observables

- [ ] Start: https://reactivex.io/documentation/operators/start.html

## Transforming Observables

- [ ] GroupBy: https://reactivex.io/documentation/operators/groupby.html
- [ ] Scan: https://reactivex.io/documentation/operators/scan.html
- [ ] Window: https://reactivex.io/documentation/operators/window.html

## Filtering Observables

- [ ] TakeLast: https://reactivex.io/documentation/operators/takelast.html
- [ ] SkipLast: https://reactivex.io/documentation/operators/skiplast.html
- [ ] Sample: https://reactivex.io/documentation/operators/sample.html
- [ ] Last: https://reactivex.io/documentation/operators/last.html
- [ ] IgnoreElements: https://reactivex.io/documentation/operators/ignoreelements.html
- [ ] First: https://reactivex.io/documentation/operators/first.html
- [ ] ElementAt: https://reactivex.io/documentation/operators/elementat.html
- [ ] Debounce: https://reactivex.io/documentation/operators/debounce.html

## Combining Observables

- [ ] And/Then/When: https://reactivex.io/documentation/operators/and-then-when.html
- [ ] CombineLatest: https://reactivex.io/documentation/operators/combinelatest.html
- [ ] Join: https://reactivex.io/documentation/operators/join.html
- [ ] StartWith: https://reactivex.io/documentation/operators/startwith.html
- [ ] Switch: https://reactivex.io/documentation/operators/switch.html

## Error Handling Operators

- [ ] Catch: https://reactivex.io/documentation/operators/catch.html
- [ ] Retry: https://reactivex.io/documentation/operators/retry.html

**Comment by Neverlord** [comment]
@shariarriday these should be straightforward to implement as a `step` (like `take`):

- First (actually, this is just `take(1)`)
- TakeLast
- Last
- IgnoreElements
- ElementAt
- SkipLast

**Comment by Neverlord** [comment]
The `catch` operator is actually more like a family of operators. For example, [RxJava](https://github.com/ReactiveX/RxJava/wiki/Error-Handling-Operators) splits this into `onErrorComplete`, `onErrorResumeNext`, `onErrorReturn`, `onErrorReturnItem` and `onExceptionResumeNext`.

We already have `on_error_complete`. We will skip the exception version, as we build our API generally around `expected`. This leaves:

- [x] `on_error_resume_next`
- [x] `on_error_return`
- [x] `on_error_return_item`

@shariarriday: I think `on_error_return` and `on_error_return_item` can be implemented as steps again. Mind giving it a shot? The function object passed to `on_error_return` should return either a `T` or an `expected<T>`. If the function object returns an `expected` with an error, the new error simply replaces the "input" error and we still call `on_error` on `next`.

**Comment by Neverlord** [comment]
`Scan` is just another name for `Reduce`, which is already implemented.

I think we can implement `start_with(x)` as `just(x).concat(*this)`. To implement `sample`, we can use the implementation of `op/buffer.hpp` as a template. Instead of having a `std::vector<input_type> buf_` member, we can simply have a `std::optional<input_type> last_` member and adjust the rest accordingly.

@shariarriday wanna give these two a shot?

**Comment by Neverlord** [comment]
@shariarriday I think `retry` is a good candidate to tackle next. Specifically, `retry_when` to retry subscribing to an input observable whenever the "control" observable emits a value. Here's a usage example:

```c++
self->make_observable()
  .fail<int>(sec::runtime_error) // subscribed to three times
  .retry_when([self](auto attempts) { // attempts: observable<error>
      return self->make_observable().zip_with(
               [](const error&, int num) { return num; },
               attempts, self->make_observable().range(1, 3))
             .flat_map([self](int num) {
               printf("delay next retry by %d second(s)", num);
               return self->make_observable().timer(std::chrono::seconds{num});
             });
  })
  .for_each([](int) { });
```

Should print:

```
delay next retry by 1 second(s)
delay next retry by 2 second(s)
delay next retry by 3 second(s)
```

And then stop after three times. With `retry_when`, we can probably build the other `retry` versions very easily later on as a followup.

**Comment by shariarriday** [comment]
Thanks, I am planning to implement `retry_when` operator after completing the implementation for `combine_latest` operator.

**Comment by dominiklohmann** [comment]
I've played a fair amount with flows lately, and I've had a need for `combine_latest` and `group_by` and `window` quite often now.

Is this issue still actively being worked on currently? I'd be willing to contribute to this (slowly over a few weekends), but would need some guidance on the implementation.

**Comment by Neverlord** [comment]
> Is this issue still actively being worked on currently?

Yes, we add new operators over time. Not with the highest priority, though.

> I've had a need for `combine_latest` and `group_by` and `window` quite often now.

Well, `window` is basically the same as `buffer` except that it emits observables instead of lists. So at least as a workaround (if you haven't already), you can emulate that by using buffer and then converting the lists to observables. Of course it'll have different timings. An implementation for `window` would probably be very similar to `buffer`, so we can probably use the latter as a template.

For `combine_latest`, @shariarriday has already started implementing it. Fell on the wayside for a bit, but we can resume work on that again knowing that there's demand for it.

Nothing resembles `group_by`, though. So that'll be the toughest one since an implementation would need to start from scratch. Also probably requires extra boilerplate code for `grouped_observable` (or some other approach to transmit the key).

> I'd be willing to contribute to this

If you're up to it: `window` might be a good candidate. There are some conceptual questions to look into first though, i.e., do the nested observables need buffering to allow "late" subscribers to read the values from the start?

**Comment by Neverlord** [comment]
After wrapping up `debounce`: we should also add `throttle_first` and `throttle_last`. Both run on a fixed interval. `throttle_first` will then emit the first observed value from that range while `throttle_last` will emit the most recent value. See https://reactivex.io/documentation/operators/sample.html. @shariarriday mind working on it?
